src/olympia/amo/middleware.py

```diff
@@ -188,15 +191,50 @@ def process_request(self, request):
 
 
 class ReadOnlyMiddleware(object):
+    """Middleware that announces a downtime which for us usually means
+    putting the site into read only mode.
+
+    Supports issuing `Retry-After` header.
+    """
+    ERROR_MSG = _(
+        u'Some features are temporarily disabled while we '
+        u'perform website maintenance. We\'ll be back to '
+        u'full capacity shortly.')
+
+    API_HEADER_NAME = 'X-AMO-Read-Only'
+
+    def _render_api_error(self):
+        response = JsonResponse({'error': self.ERROR_MSG}, status=503)
+        response[self.API_HEADER_NAME] = 'true'
+        if settings.READ_ONLY_RETRY_AFTER is not None:
+            response['Retry-After'] = settings.READ_ONLY_RETRY_AFTER.seconds
+        return response
 
     def process_request(self, request):
-        if request.method == 'POST':
+        if not settings.READ_ONLY:
+            return
+
+        if request.path.startswith('/api/'):
+            writable_method = request.method in ('POST', 'PUT', 'DELETE')
+            if settings.READ_ONLY and writable_method:
+                return self._render_api_error()
+        elif request.method == 'POST':
             return render(request, 'amo/read-only.html', status=503)
 
     def process_exception(self, request, exception):
         if isinstance(exception, mysql.OperationalError):
+            if request.path.startswith('/api/'):
+                return self._render_api_error()
             return render(request, 'amo/read-only.html', status=503)
 
+    def process_response(self, request, response):
+        # We haven't set the header yet so it's not an error response
+        header_name = self.API_HEADER_NAME
+
+        if header_name not in response and response.status_code != 503:
```

Comment:
Should this be `or` rather than `and`?

Could we just set `response[self.API_HEADER_NAME] = 'false'` in `process_request` and `process_exception` always so we don't need to do anything in `process_response`?

Reply:
> Could we just set response[self.API_HEADER_NAME] = 'false' in process_request and process_exception always so we don't need to do anything in process_response?

That's not possible so easily since I don't have access to the response in `process_request` but only in `process_response`.

That changes with new-style middlewares from Django 1.10 but we didn't convert yet.

Cleaned this up in be8dbcbe35